src/parser/core/modules/EntityStatuses.ts

```diff
@@ -0,0 +1,142 @@
+import STATUSES from 'data/STATUSES'
+import {BuffEvent} from 'fflogs'
+import Module, {dependency} from 'parser/core/Module'
+import {getDataBy} from '../../../data'
+import {FFLogsEventNormaliser} from './FFLogsEventNormaliser'
+import Invulnerability from './Invulnerability'
+
+const APPLY = 'apply'
+const REMOVE = 'remove'
+
+interface BuffTrackingEvent extends BuffEvent {
+	start: number,
+	lastRefreshed: number,
+	end: number | null,
+	stacks: number,
+	stackHistory: StackHistoryEvent[],
+	isDebuff: boolean,
+}
+
+interface StackHistoryEvent {
+	stacks: number,
+	timestamp: number,
+	invuln: boolean,
+}
+
+interface StatusRangeEndpoint {
+	timestamp: number,
+	type: 'apply' | 'remove',
+	stackHistory: StackHistoryEvent[],
+}
+
+export class EntityStatuses extends Module {
+	static handle = 'entityStatuses'
+	static debug = false
+
+	@dependency private invuln!: Invulnerability
+	@dependency private fflogsEvents!: FFLogsEventNormaliser
+
+	// -----
+	// API
+	// -----
+	// TODO: This implementation which I've shamelessly stolen seems to only track overall
+	//       uptime, ignoring potential gains from multidotting, etc. Should that be in here,
+	//       or is that a somewhere-else sort of thing...
+	getStatusUptime(statusId: number, onTargets: TODO, sourceId = this.parser.player.id) {
+		// Search for status maps on valid targets
+		const statusEvents: BuffTrackingEvent[] = Object.values(onTargets)
+			.reduce((statusEvents: BuffTrackingEvent[], target: TODO) => {
+				const appliedStatuses = target.buffs?.filter((statusEvent: BuffTrackingEvent) => {
+					return statusEvent.ability.guid === statusId && (statusEvent.sourceID === null || statusEvent.sourceID === sourceId)
+				})
+				return statusEvents.concat(appliedStatuses)
+			}, [])
+
+		let active = 0
+		let start: number | null = null
+		return statusEvents
+			.reduce((statusRanges: StatusRangeEndpoint[], statusEvent) => {
+				return statusRanges.concat(this.getStatusRangesForEvent(statusEvent))
```

Comment:
can inline the function block to avoid the `return`

Reply:
I didn't inline the function block because I was trying to separate across lines (since it's a really really long one-line otherwise) and then that causes TSLint to complain about no trailing comma (presumably due to the concat)